packages/optimizer/lib/transformers/MinifyHtml.js

```diff
@@ -0,0 +1,174 @@
+      minifyJSON: !!config.minifyJSON || true,
+      collapseWhitespace: !!config.collapseWhitespace || true,
+      removeComments: !!config.removeComments || true,
+      canCollapseWhitespace: true,
+      canTrimWhitespace: true,
+      commentIgnorePattern: config.commentIgnorePattern || COMMENT_DEFAULT_IGNORE,
+    };
+    this.log = config.log.tag('MinifyHtml');
+  }
+  transform(tree) {
+    if (!this.opts.minifyHtml) {
+      return;
+    }
+    const nodesToRemove = [];
+    this.visitNode(tree, this.opts, nodesToRemove);
+    for (const node of nodesToRemove) {
+      remove(node);
+    }
+  }
+
+  visitNode(node, opts, nodesToRemove) {
+    if (node.type === 'text') {
+      this.cleanTextNode(node, opts, nodesToRemove);
+    } else if (node.type === 'comment') {
+      this.cleanCommentNode(node, opts, nodesToRemove);
+    } else if (node.tagName === 'script') {
+      this.cleanScriptNode(node, opts);
+    }
+    const childOpts = Object.assign({}, opts);
+    if (opts.canCollapseWhitespace && !this.canCollapseWhitespace(node.tagName)) {
+      childOpts.canCollapseWhitespace = false;
+    }
+    if (opts.canTrimWhitespace && !this.canTrimWhitespace(node.tagName)) {
+      childOpts.canTrimWhitespace = false;
+    }
+    for (const child of node.children || []) {
+      this.visitNode(child, childOpts, nodesToRemove);
+    }
+  }
+
+  cleanTextNode(node, opts, nodesToRemove) {
+    if (!node.data || !opts.collapseWhitespace) {
+      return;
+    }
+    if (opts.canCollapseWhitespace) {
+      node.data = node.data.replace(WHITESPACE_REGEX, ' ');
+    }
+    if (opts.canTrimWhitespace) {
+      node.data = node.data.trim();
+    }
+    // remove empty nodes
+    if (node.data.length === 0) {
+      nodesToRemove.push(node);
+    }
+  }
+
+  cleanCommentNode(node, opts, nodesToRemove) {
+    if (!node.data || !opts.removeComments) {
+      return;
+    }
+    if (opts.commentIgnorePattern.test(node.data)) {
+      return;
+    }
+    nodesToRemove.push(node);
+  }
+
+  cleanScriptNode(node, opts) {
+    const isJson = this.isJson(node);
+    const isAmpScript = this.isInlineAmpScript(node);
+    for (const child of node.children || []) {
+      if (!child.data) {
+        continue;
+      }
+      if (isJson && opts.minifyJSON) {
+        this.minifyJson(child);
+      } else if (isAmpScript && opts.minifyAmpScript) {
+        this.minifyAmpScript(child);
+      }
+    }
+  }
+
+  minifyAmpScript(child) {
+    const result = Terser.minify(child.data);
+    if (result.error) {
+      this.log.warn(
+          'Could not minify amp-script',
+          child.data,
+          `${result.error.name}: ${result.error.message}`,
+      );
+    } else {
+      child.data = result.code;
+    }
+  }
+
+  minifyJson(child) {
+    try {
+      child.data = JSON.stringify(JSON.parse(child.data), null, '');
+    } catch (e) {
+      // invalid JSON
+      this.log.warn('Invalid JSON', child.data);
```

Comment:
If the JSON is invalid, does it still parse in the browser? Just wondering if this somehow fail more explicitly if that's the case (or maybe not in the scope of optimizer?)